fix: Allow users to handle promise rejections in ActionButton's onOk callback

grossbart opened the pull request:
Components that rely on ActionButton swallow rejected promises. This makes it impossible for userland code to handle them.

### 🤔 This is a ...

- [x] Bug fix

### 🔗 Related issue link

- Closes #40017 
- Builds on #16437

### 💡 Background and solution

Components that rely on ActionButton swallow rejected promises in their `onOk` handler. This makes it impossible for userland code to react to these errors, e.g. to log them to a remote server or to present users with an error message.

Rejected promises should not be swallowed so userland code or the browser can handle them.

<!--
1. Describe the problem and the scenario.
2. GIF or snapshot should be provided if includes UI/interactive modification.
3. How to fix the problem, and list the final API implementation and usage sample if that is a new feature.
-->

### 📝 Changelog

<!--
Describe changes from the user side, and list all potential break changes or other risks.
--->

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Allow users to handle promise rejections in ActionButton's `onOk` callback |
| 🇨🇳 Chinese | Allow users to handle promise rejections in ActionButton's `onOk` callback |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

grossbart committed:
fix: Allow users to handle promise rejections in onOk callback

Components that rely on ActionButton swallow rejected promises. This makes it impossible for userland code to handle them.

grossbart committed:
polish: Fix linting problems

afc163 commented:
<img width="1183" alt="图片" src="https://user-images.githubusercontent.com/507615/210690162-db91aa62-323f-4739-9c8c-bc18911c61e7.png">

zombieJ commented on components/_util/ActionButton.tsx:
```diff
@@ -55,12 +55,10 @@ const ActionButton: React.FC<ActionButtonProps> = (props) => {
         // See: https://github.com/ant-design/ant-design/issues/6183
         setLoading(false, true);
         clickedRef.current = false;
+        throw e;
```
Seems just return `Promise.reject` instead is much meaningful here : )

grossbart committed:
polish: Return rejected promise instead of throwing

grossbart commented:
@afc163 I've investigated why it works for me but not on CI: The problem is that as soon as Jest runs tests in parallel the proposed workaround doesn't work anymore.

My pragmatic suggestion would be to remove the test and the workaround: it just doesn't seem to be worth the trouble and in the experiments I did today I could not find a way that works across Jest processes. If you feel very strongly that a test should exist and chances are high that you would accept this PR I would investigate further.

At this moment I'm not sure a solution is possible without going deep into the internals of Jest, which would make a workaround brittle. Jest has had this issue open for four years, so I'm not sure that is on their roadmap: https://github.com/facebook/jest/issues/5620.

zombieJ commented:
Remove is not a good way. Could you help to find way out instead just remove it?

grossbart commented:
Me and a colleague dug deep into Jest's source code for several hours to see whether we could find a workaround. We could not 😞 Jest a) does not provide a real process object and b) handles `unhandledRejection` events itself, making it impossible in our understanding to change.

So, I don't see a way that this can be tested with Jest and see no other option than to remove the test …

grossbart committed:
refact: Remove test for unhandled promise rejection

This test breaks when run in parallel by Jest. At the moment we have no way of changing the way Jest works and it prohibits us from testing for "unhandledRejection" events.

See: https://github.com/ant-design/ant-design/pull/40018#issuecomment-1373590259

zombieJ commented:
Let me try other way : )

zombieJ commented:
Update. I found that this update is meanless. Nowhere to handle the return Promise.reject on `onOk`. If not care about rejection warning. Just:
https://codesandbox.io/s/antd-reproduction-template-forked-reiyjf

grossbart commented:
We're actively using this patch in production, so I know it's not meaningless 🤷 

Of course: in most situations that promise should be handled in `onOk` like you've shown. But we do not handle promise rejections in situations where no expected domain exceptions occur in the `onOk` handler. Instead we want the error to bubble up to our error boundary to handle it there. (We log it to our servers and show an error message to users.)

Without this patch, errors are swallowed and users won't know that their action didn't work. Also, as seen in #16437, it is helpful to developers to see when an error occurred. And letting an unhandled promise rejection bubble up to the browser would be the correct way to do so.

With this PR I'm trying to say: Antd should not catch the promise in library code but let developers handle it.

zombieJ committed:
test: hack for rejection

zombieJ committed:
Merge remote-tracking branch 'origin/master' into action-button-onconfirm-reject
